cmake/clang_tidy.cmake

```diff
@@ -3,10 +3,20 @@ option (ENABLE_CLANG_TIDY "Use clang-tidy static analyzer" OFF)
 
 if (ENABLE_CLANG_TIDY)
 
-    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
+    if (CLANG_TIDY_CACHE_PATH)
+        find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
+
+        # Why do we use ';' here?
```

Comment:
[CMake docs for set()](https://cmake.org/cmake/help/latest/command/set.html) say:

"Multiple arguments will be joined as a [semicolon-separated list](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-lists) to form the actual variable value to be set."

So you could just write `set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH})` and omit the comment about black magic.

Reply:
`${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH}` is not what I need.

Reply:
```
    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
    if (CLANG_TIDY_CACHE_PATH)
        set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${CLANG_TIDY_PATH})
    endif ()
```

```
CLANG_TIDY_CACHE_PATH:FILEPATH=/home/felixoid/bin/clang-tidy-cache
CLANG_TIDY_PATH:FILEPATH=/usr/bin/clang-tidy
```

It doesn't work

Reply:
And it neither

```
    find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
    find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
    if (CLANG_TIDY_CACHE_PATH)
        set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH})
    else ()
        set (CLANG_TIDY_PATH ${_CLANG_TIDY_PATH})
    endif ()
```

```
> cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=Debug -DSANITIZE= -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DENABLE_CLANG_TIDY=1 -DENABLE_TESTS=1 -DENABLE_EXAMPLES=1 -DENABLE_UTILS=1 .. | grep TIDY
By default, if no option is selected, on POSIX, libcurl transport adapter is used.
CLANG_TIDY_CACHE_PATH:FILEPATH=/home/felixoid/bin/clang-tidy-cache
ENABLE_CLANG_TIDY:BOOL=1
_CLANG_TIDY_PATH:FILEPATH=/usr/bin/clang-tidy
```

Reply:
To simplify this at least a little bit, I tried

```cmake
find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
if (CLANG_TIDY_CACHE_PATH)
    find_program (_CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
    set (CLANG_TIDY_PATH ${CLANG_TIDY_CACHE_PATH} ${_CLANG_TIDY_PATH} CACHE STRING "a doc")
else ()
    find_program (CLANG_TIDY_PATH NAMES "clang-tidy" "clang-tidy-15" "clang-tidy-14" "clang-tidy-13" "clang-tidy-12")
endif ()
```

this gives
```
By default, if no option is selected, on POSIX, libcurl transport adapter is used.
CLANG_TIDY_CACHE_PATH:FILEPATH=/home/ubuntu/bin/clang-tidy-cache                  
CLANG_TIDY_PATH:STRING=/home/ubuntu/bin/clang-tidy-cache;/usr/bin/clang-tidy      
ENABLE_CLANG_TIDY:BOOL=1                                                          
_CLANG_TIDY_PATH:FILEPATH=/usr/bin/clang-tidy                                     
```